### ISSUE TITLE
Error in userExixts

### ISSUE THREAD

**Issue by ambinabhi** [gheeres/node-activedirectory#96]
Hi
I used the activedirectory module, and i am successfully able to authenticate credentials

But when i use userExists method i m getting the following error

ERROR: {"dn":"","code":49,"name":"InvalidCredentialsError","message":"80090308:
LdapErr: DSID-0C0903A9, comment: AcceptSecurityContext error, data 52e, v1db1\u0
000"}

 i tried using the userprinciple name also , then also no luck 

Any help would be much appreciated. 

LDAP error code is 49 and description is 52e, but i am sure that both username and password which i am using are valid. I don't know what i am doing wrong.

**Comment by gheeres** [comment]
Please post sample code that is exhibiting the error. Please obfuscate any confidential information (i.e. username/password, etc.).

**Comment by ambinabhi** [comment]
var ActiveDirectory = require('activedirectory');
var config = { url: 'LDAP://xxxxx.xxxxx',
               baseDN: 'ou=Users,cn=readonly,cn=users,dc=xxxxx.xxxxx,dc=com',
               username: 'xxxxxxx',
               password: 'xxxxxxx'}

var ad = new ActiveDirectory(config);


var userPrincipalName  = 'xxxxxxx@xxxxxxxxx.com';
var username = 'xxxxxxx'
var password = 'xxxxxxx';

ad.authenticate(username, password, function(err, auth) {
  if (err) {
    console.log('ERROR: '+JSON.stringify(err));
    return;
  }

  if (auth) {
    console.log('Authenticated!');
  }
  else {
    console.log('Authentication failed!');
  }
});

ad.userExists(userPrincipalName , function(err, exists) {
  if (err) {
    console.log(userPrincipalName );
    console.log('ERROR: ' +JSON.stringify(err));
    return;
  }

  console.log(userPrincipalName  + ' exists: ' + exists);
})

**Comment by ambinabhi** [comment]
I used both username and userPrincipleName as a parameter for ad.userExists(). It still throwing up that error

**Comment by gheeres** [comment]
The ad.authenticate() method uses the username and password you provided to authenticate (bind) to the Active Directory server. It does NOT use the username and password supplied via configuration.

The ad.userExists() uses the username and password that you specified in the configuration. It does NOT user the username/password that were used in the ad.authenticate. In your config, you need to use a userPrincipalName for the username.

```javascript
var config = { url: 'LDAP://xxxxx.xxxxx',
               baseDN: 'ou=Users,cn=readonly,cn=users,dc=xxxxx.xxxxx,dc=com',
               username: 'xxxxxxx@xxxxxxxxx.com',
               password: 'xxxxxxx'}
```

In short, **EVERY** ad.*() function, except authenticate(), uses the username & password that were specified in the config object when the ad object was created to do a lookup / query operation.

**Comment by ambinabhi** [comment]
Yes i did this also, I replaced username with userPrincipleName and used both also 
I got this error 
ERROR: {"dn":"","code":49,"name":"InvalidCredentialsError","message":"80090308:
LdapErr: DSID-0C090334, comment: AcceptSecurityContext error, data 525, vece\u00
00"}

**Comment by gheeres** [comment]
Are the other ad functions (besides ad.authenticate()) giving the same exact error? If so, then you'll need to verify the userPrincipalName and password you're using in your config. The invalidCredentialsError is simply an invalid username and password and that error is being returned by the server. The other thing to check is any special characters (non ASCII) in the password. If these exists, they have to be escaped and can be difficult to get to work.

**Comment by ambinabhi** [comment]
NO i only have authenticate and userExists methods as of now. And there are special characters in the password, To which format i need to convert them ?I cant delete them to make it to work. can i ?

**Comment by gheeres** [comment]
>To which format i need to convert them ?

I don't know if that's the problem, but if other functions are also failing, that would be the likely cause. You'll need to check upstream with the [ldapjs](https://github.com/mcavage/node-ldapjs) project for information on how to escape them (if possible). The simpler solution is to use a long/complex password without any special characters.
